### ISSUE TITLE
[feature request] `get_attribute_function` option

### ISSUE THREAD

**Issue by brandonkelly** [twigphp/Twig#2338]
In Twig 2.0, the old `Twig_Template::getAttribute()` method was moved to the global `twig_get_attribute()` function, which makes it much more difficult to override than Twig 1.x, where you could just provide your own base template class with its own `getAttribute()` method.

We need to override it in Craft CMS for two reasons:

- If `$object` is an "element" class, we register the element as a dependency of any active [`cache`](https://craftcms.com/docs/templating/cache) tags, for automagic cache busting if the element is ever re-saved or deleted.

- For method calls, we convert any `Twig_Markup` objects within the `$arguments` array to strings, as Twig has access to Craft’s entire service layer, and we want to enforce strict typing on service methods whenever possible. (I submitted a [PR](https://github.com/twigphp/Twig/pull/1366) for this once but nothing came of it.)

As a workaround, we have created a custom compiler class that extends `Twig_Compiler` and overrides its `raw()` method:

```php
public function raw($string)
{
    // Use our own attribute() method
    if (strncmp($string, 'twig_get_attribute(', 19) === 0) {
        $string = 'craft\helpers\Template::attribute('.substr($string, 19);
    }

    return parent::raw($string);
}
```

Obviously this is super hacky, but it gets the job done for now. I am hoping _something_ can be added to Twig to make this better.

One possibility is a new `get_attribute_function` option, set to `'twig_get_attribute'` by default, which `Twig_Node_Expression_GetAttr::compile()` could reference rather than hardcoding `twig_get_attribute` when calling `$compiler->raw()`:

```php
public function compile(Twig_Compiler $compiler)
{
    $function = $compiler->getEnvironment()->getGetAttrFunction();
    $compiler->raw($function.'($this->env, $this->getSourceContext(), ');
    // ...
}
```

With that in place, we could remove our custom compiler class, and initialize Twig with the new option:

```php
$twig = new Twig_Environment($loader, [
    'get_attribute_function' => 'craft\helpers\Template::attribute',
]);
```

**Comment by stof** [comment]
Overriding it has never been a supported use case (btw, as soon as the C extension was installed, the method was not used anymore).

What you could do in Craft CMS is using a node visitor to replace all ``Twig_Node_Expression_GetAttr`` nodes by your own node compiled using your own function

**Comment by brandonkelly** [comment]
@stof Yeah I realize this change in v2 isn't really a regression due the C extension implication. Ironically (to me anyway) the current v2 implementation with the global `twig_get_attribute()` function would have made more sense in v1 than it does in v2.

I hadn't considered going the node visitor route, but yes that will work nicely. Thanks for the suggestion!

**Comment by stof** [comment]
@brandonkelly the move to a separate function was done initially in #2310 adding a new extension for PHP 7, to avoid having 2 different code paths. and it was merged separately to be part of 2.0 even though the merge of the extension is delayed (due to BC concerns after the release).
Twig 1.x had to change between the method and a function at compile time instead, because removing the method was not an option due to BC concerns (moving to a separate function rather than using a protected method cleaned the code btw, as ``get_class_methods`` was not usable for objects being a Twig_Template, even though we later forbid such case entirely)
